Use the least busy agent as destination

matelakat opened the pull request:
This change allows people to select which agent to use as a destination while moving routers. By default the choice was random. This patch changes the default to select always the least busy agent. The least busy agent has the smallest number of routers on it. If people want the random behaviour, they can select it with the --agent-selection-mode=random command line switch.

The new method first asks for the list of routers on each of the target agents, and saves this information in a cache. Each time when an agent is picked, the method is assuming that the move was successful, and maintains it's internal state accordingly. The cache of routers per agent will expire after 5 minutes, in which case neutron will be queried again.

matelakat commented:
![](https://github.trello.services/images/mini-trello-icon.png) [neutron-ha-tool: pick the least busy agent](https://trello.com/c/gbJcBnbO/2-neutron-ha-tool-pick-the-least-busy-agent)

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -248,6 +248,14 @@ def run(args):
+    def __init__(self, agents):
+        self.agents = agents
+
+    def pick(self):
```
nit: I would use choice here too

aspiers replied:
If so, probably `choose` is better than `choice` since it should be a verb.  But if you change it, would be better to stay consistent and rename the class to `RandomAgentChooser`.  Personally I slightly prefer `Picker` but there's not much difference.

aplanas replied:
I suggested `choice` because of the python API. I guess that `pick` is ok, even to my eyes is too similar to pickle

aplanas commented on files/default/README.md:
```diff
@@ -0,0 +1,14 @@
+
+Make a python virtual environment, install the requirements with:
+
+    mkvirtualenv hatool -p python2
```
no every one uses virtualenvwrapper, so I would suggest to use `virtualenv` command

matelakat replied:
I'm removing those steps, and just assuming the reader is capable of creating an environment.

aspiers replied:
To do this properly, isn't `tox` the best approach, consistent with upstream?

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -249,13 +264,48 @@ def run(args):
+        self.now = datetime.datetime.now
+        self.cache_created_at = None
+        self.qclient = qclient
+        self.agents_by_id = dict((agent['id'], agent) for agent in agents)
```
{agent['id']: agent for agent in agents}

matelakat replied:
nice, added

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -249,13 +264,48 @@ def run(args):
+    def refresh_router_count_per_agent_id(self):
+        LOG.info("Refreshing router count per agent cache")
+        self.router_count_per_agent_id = dict()
+        for agent_id in self.agents_by_id.keys():
```
you can drop `.keys()`

matelakat replied:
true, done

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -249,13 +264,48 @@ def run(args):
+        self.cache_created_at = None
+        self.qclient = qclient
+        self.agents_by_id = dict((agent['id'], agent) for agent in agents)
+        self.router_count_per_agent_id = dict()
```
you can ignore this, but I would use `{}`, `[]` or `()` to initialize empty dicts, lists or tuples, the reason is that is marginally faster and more usual

matelakat replied:
done

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -249,13 +264,48 @@ def run(args):
+
+    def cache_expired(self):
+        cache_life = self.now() - self.cache_created_at
+        return cache_life.total_seconds() > ROUTER_CACHE_MAX_AGE_SECONDS
```
because your measure unit is seconds, I would use `.seconds` field from the timedelta

matelakat replied:
done

aplanas commented on files/default/neutron-ha-tool.py:
```diff
@@ -249,13 +264,48 @@ def run(args):
+            self.router_count_per_agent_id.items(),
+            key=lambda x: (x[1], x[0])
+        )
+        agent_id = agent_id_number_of_routers[0][0]
```
as a safety net, check `if agent_id_number_of_routers`?

matelakat replied:
what happens here is an `IndexError` is thrown. Basically the same would happen if we used the original implementation with random:

```python
>>> import random
>>> random.choice([])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib64/python2.7/random.py", line 275, in choice
    return seq[int(self.random() * len(seq))]  # raises IndexError if seq is empty
IndexError: list index out of range
```

So I think let's leave it as it is?

aplanas replied:
We do not want to die in neutron-ha-tool, this will cause a fail in pacemaker and a migration. I would log

matelakat replied:
@aplanas the caller makes sure always that the array's length is more than one. Before we call `migrate_l3_routers_from_agent` we always check if `targets` is a non-null list. I still vote for leaving this as it is.

aplanas replied:
ok for me

aplanas commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+class TestL3AgentMigrate(unittest.TestCase):
+
+    def setUp(self):
+        import logging
```
Why the import is here? there will be not 'reimport' in every `setUp`

aplanas commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+
+    def setUp(self):
+        import logging
+        logging.basicConfig(level=logging.DEBUG)
```
this, maybe, is also not part of the setUp, can be in the global side

aplanas commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+
+    def setUp(self):
+        import logging
+        logging.basicConfig(level=logging.DEBUG)
```
same

aplanas reviewed (approved):
Basically OK from me. Except maybe the `import logger`

aspiers commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+import unittest
+import collections
+ha_tool = __import__("neutron-ha-tool")
```
That looks like a pretty unconventional way of doing the import.  Isn't it possible to do it with a normal `import`?

matelakat replied:
problem is that the filename contains dashes. I looked for other solutions, and will switch to using the `importlib` module, that might read a bit better. The good solution would be to rename the file so it does not contain dashes.

aplanas replied:
I do not know a better way neither

aspiers commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+        neutron_client = MockNeutronClient()
+        result = ha_tool.l3_agent_migrate(neutron_client)
+
+        self.assertEqual(0, result)
```
Doesn't this assume that `MockNeutronClient` starts with no agents?  That sounds a bit brittle - better to assert that as part of the test.

matelakat replied:
Yes, no agents, therefore none of them are dead, so it should not do anything. Let me amend the test, add one live agent, so it makes more sense what we are testing here. But in reality the test highlights that the tool will return 0 if we have no agents at all in the system. Whether that's a good behaviour or not - it's a different question.

aspiers commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+        neutron_client.tst_add_agent(
+            'agent-1', {
+                'agent_type': 'L3 agent',
+                'alive': False,
```
Not alive but `admin_state_up` is `True`?  That sounds strange to me but I'm probably just ignorant - in what circumstances would that happen?

matelakat replied:
The criteria for a live agent is that both `alive` and `admin_state_up` are `True`. I think I better make the `admin_state_up` to be `False` as well here. I must admit, I'm not fully aware what values could pop up here. I set one of them `False`, so that the agent will be reported as dead.

matelakat replied:
I looked at it - `alive` is reported by the agent periodically, so if the server doesn't get an `alive` message from the agent within a period of time, it will set `alive` to false. It's kind of a valid error scenario, but for now I think it's better if these initial tests reflect the happy usecases. We can add further tests to see what's the tool's response in the edge cases.

aspiers commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+        )
+        result = ha_tool.l3_agent_migrate(neutron_client)
+
+        self.assertEqual(1, result)
```
Same here, better to assert your mock's starting point.

aspiers commented on files/default/test-neutron-ha-tool.py:
```diff
@@ -0,0 +1,168 @@
+        logging.basicConfig(level=logging.DEBUG)
+
+    def test_no_agents_returns_zero(self):
+        neutron_client = MockNeutronClient()
```
Same here.

aspiers reviewed (changes requested):
Great work thanks, especially the nice refactoring of random choice into a strategy class as a separate commit before introducing a second strategy! That's exactly how to make a review task really easy and pleasant to do.  

However please don't add extra "address review comments" commits, as that makes the review process a lot harder.  Fortunately [I wrote a blog post about this](http://blog.adamspiers.org/2015/03/24/why-and-how-to-correctly-amend-github-pull-requests/) last year, so please read that.

Itxaka commented on files/default/test-requirements.txt:
```diff
@@ -0,0 +1,6 @@
+coverage
```
cant see where coverage or flake8 is being used here :)

aplanas replied:
Was documented in the readme.md, and is a good step before travis integration. IMHO it is not hurting.

Itxaka replied:
Dumb people(like me) dont read the readmes! Duh!

Disregard comment, sorry!

matelakat replied:
I think there is still a valid point, so I'll add the flake8 to the readme as well.

rhafer reviewed (approved):
Nice work! Don't have much to add to the existing  reviews. However I slightly wonder if putting the README.md and unittests into the files/default/ directory is a good choice. Or if its about time to cleanup this repository/branch and remove all the chef-specific bits from it. This is of course nothing that needs to happen as part of this PR, merely some food for thought.

We also should make sure to not merge this one before https://github.com/SUSE-Cloud/cookbook-openstack-network/pull/7 (I guess it's better to add the ```do not merge yet``` label here for now).

Commit:
Unit testing neutron-ha-tool.py

Adding some unit tests that cover neutron-ha-tool's areas where target
agents are selected.

Commit:
Refactor - extract RandomAgentPicker class

Separate the responsibility of target agent picking to it's own class.

Commit:
Refactor - add Configuration registry

Add a registry that stores the agent picking method - to make it
easier to plug in new methods for picking the target agent.

Commit:
Add LeastBusyAgentPicker

This implementation picks the least busy agent. It also maintains a
cache which is refreshed if its age is more than 5 minutes.

Commit:
Make least busy agent the default strategy

When people start the script, router placement's strategy will be the
least busy agent. Should people want to use random, they can do so via
selecting `random` for the parameter `--agent-selection-mode`.

matelakat commented:
@aspiers commits have been re-worked a bit, and I made the tests easier to read by adding some helper functions.

Itxaka commented:
small thing @matelakat , we should add *.pyc files to the .gitignore files as running the tests will generate them and we dont want someone checking them in git by mistake :)

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -867,6 +870,41 @@ def pick(self):
+
+    def cache_expired(self):
+        cache_life = self.now() - self.cache_created_at
+        return cache_life.seconds > ROUTER_CACHE_MAX_AGE_SECONDS
```
Not sure if it really matters, but strictly speaking, I guess having a `cache_life.days > 0` check wouldn't hurt.

matelakat replied:
Right, good spot. Look at this:
```python
>>> import datetime
>>> now =datetime.datetime.now()
>>> tomorrow = now + datetime.timedelta(days=1)
>>> (tomorrow - now).seconds
0
>>> (tomorrow - now).total_seconds()
86400.0
```

I guess I'll use `total_seconds` as I did initially <- @aplanas

aplanas replied:
Oh I see! I was wrong here and `.total_seconds` is correct.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -753,24 +847,75 @@ def target_agent_list(agent_list, agent_type, exclude_agent_host):
+        )
+        agent_id = agent_id_number_of_routers[0][0]
+        self.router_count_per_agent_id[agent_id] += 1
+        return self.agents_by_id[agent_id]
```
Shouldn't we check that the agent is still alive? (it's actually not specific to this class, was also true for the random case) I can imagine the case where an agent dies, neutron-ha-tool is called to migrate things synchronously, but because it's synchronous, it takes some time and a second agent dies... Or is it purely theoretical?

matelakat replied:
I think you have a valid point there @vuntz , but I technically see that as a separate feature. What I'm thinking about is adding an agent checker that will check the agents according to some method. That would mean that the agent pickers would need to support on-the fly agent removal.

Let's see what happens now: I believe when we poll the port status, we will time out for that router and for all the routers that were targeted to the same agent. So our actual response is timing out horribly, and on the next run the dead agent is detected. This is a bit far from optimal. I think we would need to address it, as we'll have higher timeouts already in the system due to this synchronous nature of router movement.

A fail-fast approach: when polling the ports, also poll the agent for it's liveness. When we find that id died, we raise. Next run will take care of it.

So another approach might be: When we poll the ports, we should also poll the agent status. If we find that the agent has died in the meanwhile, that triggers:
 - Put the actual router back to the queue
 - Re-list the live agents

Let me create a separate card for this item.

aspiers replied:
Very good discussion here, thanks guys :)

matelakat commented:
@aspiers Let me know if I have the green light on this.

vuntz commented on files/default/neutron-ha-tool.py:
```diff
@@ -104,14 +107,19 @@ def parse_args():
+                    default='least-busy',
+                    help='Determines how target agent is selected for routers '
+                         '"random" selects target agent randomly, '
+                         '"least-busy" selects the least busy agent.')
```
It sounds to me like the output of this may not be the three lines you have here -- couldn't it happen that the first two lines are merged, which will result in some not so nice text?

aspiers replied:
@vuntz Nope, the `argparse` module automatically does line-wrapping by default, although I admit that this is quite unclear from the docs:

https://docs.python.org/2/library/argparse.html#formatter-class

vuntz reviewed (commented):
You may consider this as +1; just one comment where I'm not 100% sure that I'm correct.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -106,12 +106,12 @@ def parse_args():
     wait_parser = ap.add_mutually_exclusive_group(required=False)
     wait_parser.add_argument('--wait-for-router', action='store_true',
-                    dest='wait_for_router')
+                             dest='wait_for_router')
```
flake8 is separate to unit testing so should have been in a separate commit.  I was initially confused by trying to figure out how these whitespace changes were related to unit testing, before realising that they are not.

matelakat replied:
You are right, they should have been in separate commits - Let me split them - although to be fair, the commit message said that it contains both flake8 and unit testing.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -764,7 +764,7 @@ def list_routers(qclient):
     LOG.debug("list_routers: %s", resp)
     # Filter routers to not include HA routers
-    return [i for i in resp['routers'] if not i.get('ha') == True]
+    return [i for i in resp['routers'] if not i.get('ha') == True]  # noqa
```
What's this `noqa` comment for?

matelakat replied:
To silence the flake8 violation on this line: `E712 comparison to True should be 'if cond is True:' or 'if cond:'`. I wasn't brave enough to fix the violation itself, as I see that as a separate change, so for now to silence the violation, I added the `#noqa`, see https://pypi.python.org/pypi/flake8 .

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -776,7 +776,7 @@ def list_routers_on_l3_agent(qclient, agent_id):
     resp = qclient.list_routers_on_l3_agent(agent_id)
     LOG.debug("list_routers_on_l3_agent: %s", resp)
-    return [r['id'] for r in resp['routers'] if not r.get('ha') == True]
+    return [r['id'] for r in resp['routers'] if not r.get('ha') == True]  # noqa
```
Same question here.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -913,6 +922,14 @@ def pick(self):
+    Registry for storing application's configuration
+    """
+
+    agent_picker_class = RandomAgentPicker
```
Hmm.  This isn't really much more than a glorified global variable which just happens to have been wrapped inside a class, and in general use of globals should be treated with caution.  It lets you avoid passing the configuration around via method calls, but that results in the usage scope not being clearly defined, and other disadvantages associated with globals.  I won't give a -1 since it probably doesn't cause any real harm here, but I'd recommend thinking twice before using this design pattern in the future.  http://wiki.c2.com/?GlobalVariablesConsideredHarmful and http://wiki.c2.com/?GlobalVariablesAreBad are interesting reading ...

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -23,6 +23,7 @@
 
 import argparse
```
GitHub doesn't let me review the commit message directly, but it should say "its" not "it's" ;-)  "it's" always means "it is".

matelakat replied:
> GitHub doesn't let me review the commit message directly, but it should say "its" not "it's" ;-) "it's" always means "it is".

Thanks, will amend it.

aspiers commented on files/default/neutron-ha-tool.py:
```diff
@@ -922,6 +925,41 @@ def pick(self):
 
+class LeastBusyAgentPicker(object):
+    def __init__(self, qclient, agents):
+        self.now = datetime.datetime.now
```
Is it good practice for a class to "steal" methods from another class like this?  I don't know, but I'm dubious it is, because it changes the context in which the method is executed.  If the method relies on internal state (which it probably doesn't in this case) then couldn't that be evaluated in the wrong context?  I would have thought that

    def now(self):
        return datetime.datetime.now()

would be safer and less voodoo, but I'd be happy to hear that I'm wrong.

matelakat replied:
The intention of having now being injectable was because I wanted to have the ability to replace `now` with a mock should we need. Looking at the tests, this was never used, so it's a candidate for YAGNI now. At a later change this has been removed also. I might move that change here, so we don't do the add-and-remove cycle.

aspiers reviewed (approved):
Great work!  Still a few opportunities to polish this further if you are feeling diligent, but it's probably already good enough to merge.  I'd appreciate responses to my questions either way, though.

vuntz commented:
@matelakat not sure if there are still changes pending here? If not, I guess you can merge since you got 4 approvals :-)

matelakat commented:
@vuntz - @aspiers had some comments. For some reason they are only visible when you review the change commit-by commit.

Commit:
Add Code Analysis

Add flake8, fix whitespace violations, ignore others for now.
